src/cost_estimates/pessimistic.rs

```diff
@@ -0,0 +1,297 @@
+use util::db::u64_to_sql;
+use vm::costs::ExecutionCost;
+
+use core::BLOCK_LIMIT_MAINNET;
+
+use crate::util::db::tx_begin_immediate_sqlite;
+
+use super::{CostEstimator, EstimatorError};
+
+/// This struct pessimistically estimates the `ExecutionCost` of transaction payloads.
+///
+/// Each operation has a string-valued key (see `PessimisticEstimator::get_estimate_key`).
+///
+/// For each pair of 1) operation key, and 2) dimension of
+/// ExecutionCost, the PessimisticEstimator retains a set of the top
+/// 10 highest costs yet observed for that operation/dimension. The
+/// estimate returned is the average of these.
+pub struct PessimisticEstimator {
+    db: Connection,
+    log_error: bool,
+}
+
+#[derive(Debug)]
+struct Samples {
+    items: Vec<u64>,
+}
+
+const SAMPLE_SIZE: usize = 10;
+const CREATE_TABLE: &'static str = "
+CREATE TABLE pessimistic_estimator (
+    estimate_key TEXT PRIMARY KEY,
+    current_value NUMBER NOT NULL,
+    samples TEXT NOT NULL
+)";
+
+iterable_enum!(CostField {
+    RuntimeCost,
+    WriteLength,
+    WriteCount,
+    ReadLength,
+    ReadCount,
+});
+
+impl CostField {
+    /// Select `self` out of the given ExecutionCost
+    fn select_key(&self, from_cost: &ExecutionCost) -> u64 {
+        match self {
+            CostField::RuntimeCost => from_cost.runtime,
+            CostField::WriteLength => from_cost.write_length,
+            CostField::WriteCount => from_cost.write_count,
+            CostField::ReadLength => from_cost.read_length,
+            CostField::ReadCount => from_cost.read_count,
+        }
+    }
+}
+
+impl std::fmt::Display for CostField {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CostField::RuntimeCost => write!(f, "runtime"),
+            CostField::WriteLength => write!(f, "write-length"),
+            CostField::WriteCount => write!(f, "write-count"),
+            CostField::ReadLength => write!(f, "read-length"),
+            CostField::ReadCount => write!(f, "read-count"),
+        }
+    }
+}
+
+impl FromSql for Samples {
+    fn column_result(
+        sql_value: rusqlite::types::ValueRef<'_>,
+    ) -> rusqlite::types::FromSqlResult<Self> {
+        let json_value = JsonValue::column_result(sql_value)?;
+        let items = serde_json::from_value(json_value).map_err(|_e| {
+            error!("Failed to parse PessimisticEstimator sample from SQL");
+            FromSqlError::InvalidType
+        })?;
+        Ok(Samples { items })
+    }
+}
+
+impl Samples {
+    fn to_json(&self) -> JsonValue {
+        JsonValue::from(self.items.as_slice())
+    }
+
+    /// Add a new sample to this struct. The pessimistic sampler only adds to the sample set
+    ///  if the sample set is less than SAMPLE_SIZE or the new sample is greater than the current min.
+    /// Boolean return indicates whether or not the sample was included.
+    fn update_with(&mut self, sample: u64) -> bool {
```

Comment:
Just to be clear - there's no way for the sample values to get smaller?  So, if we deploy bigger and bigger smart contracts, for example, the cost estimate for deploying them will continue to get bigger and bigger (given the `mean()` calculation below)?

Reply:
Yeah, that's true for smart contract deployment in this naively pessimistic estimator --- smart contract deployment will always be estimated as pretty costly, and perhaps a linear model that takes into account contract size would work better for deploy type transactions. However contract-call's are estimated independently, so this isn't the case for those kinds of transactions.